Fixes for importing still images with an image_range

graeme-winter opened the pull request:
Fixes for importing still images with an image_range i.e.

```
dials.import path/to/grid.h5 image_range=100,200
```

Resolves #1592, includes test to prevent regression.

dagewa commented:
I'd like to have a look to see what happens with scans, re `subset = imageset[j - offset : j - offset + 1]`. Should be something similar, right?

rjgildea commented on test/command_line/test_import.py:
```diff
@@ -474,6 +471,36 @@ def test_import_still_sequence_as_experiments_subset(dials_data, tmpdir):
+
+    iset = set(exp.imageset for exp in imported_exp)
+    assert len(iset) == 1
+    assert len(imported_exp[0].imageset) == 3
```
My one concern with this test is that it only tests that we only import 3 images, not which 3 images are imported. I.e. we don't explicitly check that we have imported images 3-5 rather than 1-3 (just thinking of ways in which the code could break, yet the tests still pass).

Perhaps checking the summed counts for each image in the imported imageset would suffice (assuming that each image has different total counts)?

graeme-winter replied:
This point is a good point

graeme-winter replied:
```
assert list(iset)[0].get_image_identifier(0) == image_files[2].strpath
```

seemed simpler

rjgildea commented on command_line/dials_import.py:
```diff
@@ -571,9 +571,19 @@ def __call__(self, imageset_list):
+                    # check if user has overridden the input - if yes, recall
+                    # that these are in people numbers (1...) and are inclusive
+                    if self.params.geometry.scan.image_range is not None:
+                        ustart, uend = self.params.geometry.scan.image_range
```
```suggestion
                        user_start, user_end = self.params.geometry.scan.image_range
```
I assume "u" is short for "user" here? Might be worth spelling this out long hand to save future me the trouble of trying to figure out "what on earth does 'u' stand for?".

dagewa commented:
> I'd like to have a look to see what happens with scans, re `subset = imageset[j - offset : j - offset + 1]`. Should be something similar, right?

Ok, I chased this through `ExperimentListTemplateImporter`  --> `DataBlockTemplateImporter` --> `ImageSetFactory.make_sequence` and I'm lost and don't really know why it didn't just work already. The first first and last images should get set in the `Scan` object, and we should be able to use this to get the image range even if this is a still series

graeme-winter commented:
> > I'd like to have a look to see what happens with scans, re `subset = imageset[j - offset : j - offset + 1]`. Should be something similar, right?
> 
> Ok, I chased this through `ExperimentListTemplateImporter` --> `DataBlockTemplateImporter` --> `ImageSetFactory.make_sequence` and I'm lost and don't really know why it didn't just work already. The first first and last images are set in the `Scan` object, and we use that to access the images we want, even if it is a stills series. I think I agree with the assessment that this adds technical debt, because there is something more fundamental broken.

It's different, as a still sequence is N experiments with N scans, whereas rotation data are 1 experiment with 1 scan

rjgildea commented on command_line/dials_import.py:
```diff
@@ -571,9 +571,19 @@ def __call__(self, imageset_list):
+                    if self.params.geometry.scan.image_range is not None:
+                        user_start, user_end = self.params.geometry.scan.image_range
+                        offset = imageset.get_scan().get_array_range()[0]
+                        start, end = ustart - 1, uend
```
```suggestion
                        start, end = user_start - 1, user_end
```
Sorry, the above suggested change was incomplete (I should have modified it fully such that applying the suggested change would work).

graeme-winter commented on test/command_line/test_import.py:
```diff
@@ -474,6 +471,36 @@ def test_import_still_sequence_as_experiments_subset(dials_data, tmpdir):
+
+    out = "experiments_as_still.expt"
+
+    _ = procrunner.run(
```
Done

graeme-winter commented on test/command_line/test_import.py:
```diff
@@ -474,6 +471,36 @@ def test_import_still_sequence_as_experiments_subset(dials_data, tmpdir):
+            "output.experiments=%s" % out,
+        ]
+        + [f.strpath for f in image_files],
+        working_directory=tmpdir.strpath,
```
🤷‍♂️

graeme-winter replied:
@ndevenish added changes

ndevenish commented on command_line/dials_import.py:
```diff
@@ -571,9 +571,19 @@ def __call__(self, imageset_list):
+
+                    # check if user has overridden the input - if yes, recall
+                    # that these are in people numbers (1...) and are inclusive
+                    if self.params.geometry.scan.image_range is not None:
```
```suggestion
                    if self.params.geometry.scan.image_range:
```
Are there any non-None valid falseish settings for this parameter?

ndevenish commented on test/command_line/test_import.py:
```diff
@@ -474,6 +471,38 @@ def test_import_still_sequence_as_experiments_subset(dials_data, tmpdir):
+
+    assert result.returncode == 0
+
+    imported_exp = load.experiment_list(str(out))
```
```suggestion
    imported_exp = load.experiment_list(out)
```
`load.experiment_list` [explicitly](https://github.com/cctbx/dxtbx/blob/daf2542a15134ddb54c03a4af43a80335a5d735a/serialize/load.py#L119-L122) supports the `pathlib` protocols

ndevenish commented on test/command_line/test_import.py:
```diff
@@ -474,6 +471,38 @@ def test_import_still_sequence_as_experiments_subset(dials_data, tmpdir):
+            "dials.import",
+            "scan.oscillation=10,0",
+            "image_range=3,5",
+            "output.experiments=%s" % out,
+        ]
+        + [f.strpath for f in image_files],
```
```suggestion
            f"output.experiments={out}",
            *image_files,
        ],
```
I find it really hard to notice that the comprehension is part of the build. I checked, and procrunner/subprocess appears to deal fine with both `py.path` and `pathlib` objects.

ndevenish reviewed (approved):
I'm not entirely sure that I can follow the start/end/offset shifting. Would feel happier with Richard's suggestion of confirming we got the right range and that we haven't e.g. just introduced _another_ off-by-one error somewhere

graeme-winter commented:
> I'm not entirely sure that I can follow the start/end/offset shifting. Would feel happier with Richard's suggestion of confirming we got the right range and that we haven't e.g. just introduced _another_ off-by-one error somewhere

Added an assertion that the first image in the image set is the image we _think_ it should be.

It is also (FWIW) the same image as importing the numbered images as elsewhere in the tests.

ndevenish commented:
I will merge this when the tests pass 👍
